core/templates/dev/head/components/question-directives/question-player/services/question-player-state.service.spec.ts

```diff
@@ -0,0 +1,93 @@
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the question player state service.
+ */
+
+require(
+  'components/question-directives/question-player/services/' +
+  'question-player-state.service.ts');
+
+describe('Question player state service', function() {
+  var qpservice;
+  var questionId = 'question_1';
+
+  beforeEach(angular.mock.module('oppia'));
+
+  beforeEach(angular.mock.inject(function($injector) {
+    qpservice = $injector.get('QuestionPlayerStateService');
+  }));
+
+  it('should return an empty question state dictionary', function() {
+    expect(qpservice.getQuestionStateData()).toEqual({});
+  });
+
+  it('should add a hint to the question state data', function() {
+    qpservice.hintUsed(questionId);
+    var stateData = qpservice.getQuestionStateData();
+    expect(stateData[questionId]).toBeTruthy();
+    expect(stateData[questionId].usedHints).toBeDefined();
+    expect(stateData[questionId].usedHints.length).toEqual(1);
+    expect(stateData[questionId].usedHints[0].timestamp).toBeDefined();
+    expect(stateData[questionId].usedHints[0].timestamp).toBeGreaterThan(0);
+  });
+
+  it('should record a wrong answer was submitted to the question state data',
+    function() {
+      qpservice.answerSubmitted(questionId, false);
+      var stateData = qpservice.getQuestionStateData();
+      expect(stateData[questionId]).toBeTruthy();
+      expect(stateData[questionId].answers).toBeDefined();
+      expect(stateData[questionId].answers.length).toEqual(1);
+      expect(stateData[questionId].answers[0].isCorrect).toEqual(false);
+      expect(stateData[questionId].answers[0].timestamp).toBeGreaterThan(0);
+    });
+
+  it('should record a right answer was submitted to the question state data',
+    function() {
+      qpservice.answerSubmitted(questionId, true);
+      var stateData = qpservice.getQuestionStateData();
+      expect(stateData[questionId]).toBeTruthy();
+      expect(stateData[questionId].answers).toBeDefined();
+      expect(stateData[questionId].answers.length).toEqual(1);
+      expect(stateData[questionId].answers[0].isCorrect).toEqual(true);
+      expect(stateData[questionId].answers[0].timestamp).toBeGreaterThan(0);
+    });
+
+  it('should record that a solution was viewed',
+    function() {
+      qpservice.solutionViewed(questionId);
+      var stateData = qpservice.getQuestionStateData();
+      expect(stateData[questionId]).toBeTruthy();
+      expect(stateData[questionId].viewedSolution).toBeDefined();
+      expect(stateData[questionId].viewedSolution.timestamp).toBeDefined();
+      expect(
+        stateData[questionId].viewedSolution.timestamp).toBeGreaterThan(0);
+    });
+
+  it('should shouldn\'t record a correct answer if a solution was viewed',
```

Comment:
Maybe rename this test to 'record any answer' or 'record last answer submitted' as answers array can have multiple answers, just that the last one won't be counted if solution was viewed, right?

Reply:
It will still record answers. It just shouldn't record a correct answer if the solution was viewed. (Because we shouldn't count the score of a correct answer since the user viewed the solution) See line 53 in question-player-state.service where this is done.

Reply:
Oh, ok, yeah, fine then.